### ISSUE TITLE
Add an option to the dotnet host to enable app-crashing exception details to be printed on a single line

### ISSUE THREAD

**Issue by DamianEdwards** [dotnet/runtime#92363]
# Problem

Some environments where apps are running, e.g. Azure App Service/Container Apps, are configured to capture the console output and egress it to a table-based storage mechanism. In these circumstances, the new line character is used as the separator between individual log messages. In situations where an application is crashing on startup, before the app's usual logging system is initialized, perhaps due to a misconfiguration, the exception detail message printed by the `dotnet` host will result in many log records being created in the storage system, often interleaved with other messages from other instances of the application. This can make it very difficult to read the full exception details and diagnose why the application is crashing.

Here's a simple example of what how this looks today for a console app that crashes with a callstack a few frames deep. Note that many apps often have much deeper callstacks when failing at startup using our frameworks:

*Program.cs*

```csharp
ConfigureMe();

Console.WriteLine("Hello, World!");

void ConfigureMe()
{
    SomeNestedItemBeingConfigured();
}

void SomeNestedItemBeingConfigured()
{
    throw new InvalidOperationException("Oh no, I'm misconfigured!!\r\nPlease ensure a value for the 'SpecialValueNeeded' configuration key is defined.");
}
```

*Console output*

```
$ .\bin\Debug\net8.0\CrashyMcCrashFace.exe
Unhandled exception. System.InvalidOperationException: Oh no, I'm misconfigured!!
Please ensure a value for the 'SpecialValueNeeded' configuration key is defined.
   at Program.<<Main>$>g__SomeNestedItemBeingConfigured|0_1() in D:\src\local\CrashyMcCrashFace\Program.cs:line 12
   at Program.<<Main>$>g__ConfigureMe|0_0() in D:\src\local\CrashyMcCrashFace\Program.cs:line 7
   at Program.<Main>$(String[] args) in D:\src\local\CrashyMcCrashFace\Program.cs:line 1
```

*Output captured in log tables*

<img width="593" alt="image" src="https://github.com/dotnet/runtime/assets/249088/49a40936-374f-41d7-99a9-d91afc126087">

# Proposal

A runtime host option to force application-crashing exception details to be printed on a single line would improve the experience in these environments. The option would be settable via an environment variable, _ApplicationName.runtimeconfig.json_ JSON file, or property in the project file. An option to support JSON formatting the output could be interesting too.

Perhaps the option could be named something like `DOTNET_CRASH_EXCEPTION_FORMAT` with values of:

- `MULTI_LINE`
    Default, does what it does today
- `SINGLE_LINE`
    Replaces all new line chars with inline whitespace
- `JSON`
    Produces a single line JSON string containing similar details to what the host logs to the Windows Application Event Log on app crash today, e.g.:
    ```json
    {"faultingApplication":{"name":"CrashyMcCrashFace.exe","version":"1.0.0.0","timeStamp":"2023-09-20T18:25:43.511Z"},"exceptionCode":"0xe0434352","faultOffset":"0x000000000006531c","faultingProcessId":"1234","faultingApplicationStartTime":"2023-09-20T18:25:43.491Z","faultingApplicationPath":"D:\\src\\local\\CrashyMcCrashFace\\bin\\Debug\net8.0\\CrashyMcCrashFace.exe"}
    ```

**Comment by jkotas** [comment]
> Some environments where apps are running, e.g. Azure App Service/Container Apps, are configured to capture the console output and egress it to a table-based storage mechanism

Does this table-based storage mechanism support unlimited single line lengths?

**Comment by DamianEdwards** [comment]
> > Some environments where apps are running, e.g. Azure App Service/Container Apps, are configured to capture the console output and egress it to a table-based storage mechanism

> Does this table-based storage mechanism support unlimited single line lengths?

That's a very good question. From a [quick search online](https://github.com/MicrosoftDocs/azure-docs/issues/55225#issuecomment-631193693) the max length is 32 KB. Longer than that and it will be truncated.

**Comment by jkotas** [comment]
It is very common for error messages to span multiple lines. Here is an example of error message for Python: https://learn.microsoft.com/en-us/azure/container-instances/container-instances-get-logs#view-logs

Is this special option going to negatively impact output readability of `az container logs` output and other similar places that assume that the output is human friendly?

Is every language out there expected to implement a single-line error formatting option in order to play well with Azure App Service/Container Apps?

**Comment by DamianEdwards** [comment]
Agreed on all points. Like many things, there are trade-offs involved. This would be opt-in of course. Having the platform host capture console logs in their "natural" format would always be preferred if possible, but this was raised as a possible mitigation.
